Add Instana exporter overall structure

hickeyma opened the pull request:
Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Cedric Ziel <cedric@cedric-ziel.com>
Co-authored-by: Willian Carvalho <willianc@pm.me>

**Description:** 

Adding new component, Instana exporter. 

This is the first of 3 PRs which adds the overall skeleton structure.

**Link to tracking Issue:** 

Partial #13395 

**Testing:** 

Unit Tests

**Documentation:**

Readme explaining the exporter and its configuration.

hickeyma committed:
Add Instana exporter overall structure

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Martin Hickey <martin.hickey@ie.ibm.com>
Co-authored-by: Cedric Ziel <cedric@cedric-ziel.com>
Co-authored-by: Willian Carvalho <willianc@pm.me>

hickeyma commented:
/assign @jpkrohling

hickeyma commented:
/cc @jpkrohling

jpkrohling commented on exporter/instanaexporter/README.md:
```diff
@@ -0,0 +1,73 @@
+
+| Status                   |                  |
+| ------------------------ |------------------|
+| Stability                | [beta]           |
```
Please, review the possible values and select accordingly. For this PR, it should probably be "in development", switching to "alpha" once the component is ready. "beta" implies that this component has had at least "some" exposure to production workloads.

hickeyma replied:
Agree, updated.

jpkrohling commented on exporter/instanaexporter/config/config.go:
```diff
@@ -0,0 +1,64 @@
+const (
+	// AttributeInstanaHostID can be used to distinguish multiple hosts' data
+	// being processed by a single collector (in a chained scenario)
+	AttributeInstanaHostID = "instana.host.id"
```
Do you really mean them all to be exported?

hickeyma replied:
Good catch, moved to internal.

jpkrohling commented on exporter/instanaexporter/factory.go:
```diff
@@ -0,0 +1,88 @@
+	// The value of "type" key in configuration.
+	typeStr = "instana"
+	// The stability level of the exporter.
+	stability = component.StabilityLevelBeta
```
In development

hickeyma replied:
Done

jpkrohling commented on exporter/instanaexporter/factory.go:
```diff
@@ -0,0 +1,88 @@
+			Endpoint: "",
+			Timeout:  30 * time.Second,
+			Headers:  map[string]string{},
+			// We almost read 0 bytes, so no need to tune ReadBufferSize.
```
Not sure I get this comment.

hickeyma replied:
Agree, removed

jpkrohling commented on exporter/instanaexporter/factory.go:
```diff
@@ -0,0 +1,88 @@
+		ctx,
+		set,
+		config,
+		instanaExporter.pushConvertedTraces,
```
Where is this one?

hickeyma replied:
It is part of the Instana implementation code. This is pushed in the 2nd PR.

jpkrohling reviewed (commented):
This PR seems to be missing a few things from the skeleton, like the Makefile.

hickeyma committed:
Update after reviews

Review comments:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13452#pullrequestreview-1080778360

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
@jpkrohling Thanks for the review and feedback.

I have updated your comments and it is ready for review again.

> This PR seems to be missing a few things from the skeleton, like the Makefile

I am following the [When adding a new component](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component) and for 1st PR it suggests the: "Readme, configuration, and factory implementation". The Makefile etc. can be added in 2nd PR with the implementation, so it can build.

jpkrohling commented:
> The Makefile etc. can be added in 2nd PR with the implementation, so it can build.

Interesting, so far, our interpretation has been that the Makefile has to be part of the first PR so that it's incorporated into the main build. This way, we can ensure the files are syntactically correct, like the `config.go`. I believe the only thing you'd have to change in this PR is remove the reference to the non-existing function from the factory.

hickeyma committed:
Update after review

Review comments:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13452#issuecomment-1222832369

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented:
Thanks @jpkrohling for the feedback. I have added the Makefile as you suggested. Let me know if it needs any other changes.

hickeyma committed:
Integrate exporter into the components

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update changelog

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Add porto changes

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

jpkrohling commented on .github/CODEOWNERS:
```diff
@@ -45,6 +45,7 @@ exporter/googlemanagedprometheusexporter/            @open-telemetry/collector-c
 exporter/googlecloudpubsubexporter/                  @open-telemetry/collector-contrib-approvers @alexvanboxel
 exporter/humioexporter/                              @open-telemetry/collector-contrib-approvers @xitric
 exporter/influxdbexporter/                           @open-telemetry/collector-contrib-approvers @jacobmarble
+exporter/instanaexporter/                            @open-telemetry/collector-contrib-approvers @hickeyma
```
Given I'm the sponsor, I believe I should be part of the code owners for this component.

jpkrohling commented on exporter/instanaexporter/README.md:
```diff
@@ -0,0 +1,73 @@
+
+| Status                   |                  |
+| ------------------------ |------------------|
+| Stability                | [in-development]  |
```
Not for this PR, but the table deserves properly formatting :-)

hickeyma replied:
Its is unaligned here in the comparison but it is aligned when the markdown is rendered.

hickeyma committed:
unit tests

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma committed:
Update after review

Review comments:
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13452#discussion_r952745336
- https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/13452#discussion_r952748104

Signed-off-by: Martin Hickey <martin.hickey@ie.ibm.com>

hickeyma commented on exporter/instanaexporter/factory.go:
```diff
@@ -0,0 +1,92 @@
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/exporter/exporterhelper"
+
+	instanaConfig "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/instanaexporter/config"
```
Just a throw back when internal constants were in the config. Moved it to the root now.

hickeyma commented:
@jpkrohling Thanks for the reviews and feedback. I have updated it and it should be good for review again.

hickeyma commented:
Thanks @jpkrohling for review, feedback and merging.
